### ISSUE TITLE
Invoice Grand Total wrong if discount applied

### ISSUE THREAD

**Issue by HenKun** [magento/magento2#32270]
### Preconditions (*)
1. Magento 2.4.1
2. `tax/calculation/price_includes_tax` => `1`
3. `tax/calculation/apply_after_discount` => `1`
4. `tax/calculation/discount_tax` => `1`

### Steps to reproduce (*)

1. Create a product with price 20.50 (including tax as per setting above)
2. Tax is set to 19%
3. Create a cart rule for 100% discount
4. Order the product
5. Create invoice

### Expected result (*)

1. Order Grand Total is 0
2. Order Tax is 0
3. Invoice Tax is 0
4. Invoice Grand Total excl. Tax is 0
5. Invoice Grand Total incl.  Tax is 0

### Actual result (*)

1. Order Grand Total is 0
2. Order Tax is 0
3. Invoice Tax is 0
4. Invoice Grand Total excl. Tax is 3.27
5. Invoice Grand Total incl.  Tax is 3.27

### Probable Reason

In `vendor/magento/module-sales/Model/Order/Invoice/Total/Tax.php` on line 121 the invoice grand total is set to:

```
$invoice->setGrandTotal($invoice->getGrandTotal() + $totalTax + $totalDiscountTaxCompensation);
```

which resolves to:

```
$invoice->setGrandTotal(0 + 0 + 3.27);
```

The usage of `$totalDiscountTaxCompensation` is IMHO wrong when using these settings. 

We now have the scenario that the invoices show a grand total > 0 which is not resolvable for accounting since these amounts were never received from the customer.

Basically the same issue occurs if the discount is not 100%, just the amounts differ of course.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by HenKun** [comment]
@magento give me 2.4-develop instance

**Comment by HenKun** [comment]
Could be reproduced on test instance:

Invoice:
![image](https://user-images.githubusercontent.com/18382633/108994412-0e30ff80-769c-11eb-8a14-f3ccebbc60d0.png)

Order:
![image](https://user-images.githubusercontent.com/18382633/108994488-256fed00-769c-11eb-98cb-3e5ba6ddaf24.png)

**Comment by willryanuk** [comment]
I'm seeing the same thing on Magento 2.4.2 CE.

**Comment by HenKun** [comment]
@willryanuk I don't think these tickets are directly related. The tax is basically calculated correctly (against the assumptions in the linked tickets). As you see in your and my examples, the tax is 0 which is correct.

However, the invoice grand total contains the hidden tax (which is the tax that would be applied without the discount) that results in

1) the grand total of invoice is greater than 0
2) the total paid amount shown in order is greater than 0

whereas the customer actually (and correctly) paid nothing since the order grand total is 0.

So in my opinion the handling or application of the hidden tax is incorrect.

And this is a great issue for us at the moment.

**Comment by HenKun** [comment]
@willryanuk are you sure smaller discounts work fine? I think the problem ist just not that obvious

**Comment by willryanuk** [comment]
@HenKun unfortunately (and confusingly) it does indeed seem to invoice correctly for smaller discounts. 

E.g. here is a discount at 10% where the discount is calculated correctly, and the total paid is calculated correctly. So it only seems to happen for me for larger discounts such as 100%.

![image](https://user-images.githubusercontent.com/374656/109508645-fa204000-7a97-11eb-895b-81397a1c9a2d.png)

**Comment by HenKun** [comment]
Indeed, that seems to be correct. Since you seem to have a test setup at hand, can you check if a 99% discount gives correct or wrong results?

**Comment by willryanuk** [comment]
Yes, a very high discount e.g. 99% gives erroneous totals. Here is a summary from an order that I performed last week which illustrates the bug...

![image](https://user-images.githubusercontent.com/374656/109511649-2a1d1280-7a9b-11eb-8f5f-e8eb6027ca4d.png)

**Comment by HenKun** [comment]
OK, that's really weird. Let's wait if a Magento Developer has more insight into the problem.

**Comment by JoostWan** [comment]
Same problem over here with a coupon with 100% discount on Magento 2.3.6. Total Paid amount is not zero but the amount is the tax of the order.

**Comment by barryvdh** [comment]
This is probably related to https://github.com/magento/magento2/issues/30853

**Comment by HenKun** [comment]
@barryvdh Seems to be the same issue, correct. Especially the comment:
> I might guess that this is of big concerns when discount is greater than 50%, as, if the GrandTotal gest negative along the overall computation, a recent Magento update f**s it up.

indicates why the large amount of 99% fails and the discount of 10% succeeds.

Even more sad that this bug is present since 2.3.6 and known since Nov 2020 and not fixed yet.

**Comment by JoostWan** [comment]
@HenKun Fixed my issue (100% discount) with the following change https://github.com/magento/magento2/pull/31556/commits/34ecd22341ba95e13c6dd7407dd74d6b20318487

**Comment by willryanuk** [comment]
Fixed it for me as well. I waste a lot of time on Magento with bugs such as this.

**Comment by HenKun** [comment]
@JoostWan @willryanuk 
I applied the patch of https://github.com/magento/magento2/commit/34ecd22341ba95e13c6dd7407dd74d6b20318487 which is the same as PR https://github.com/magento/magento2/pull/31556

However, that does not work for me for discounts smaller than 100%. Given a discount of 90%, the invoice grand total is just 0.

Can you check if this is only on my side?

![grafik](https://user-images.githubusercontent.com/18382633/110608290-667ffb00-818c-11eb-807e-159310de6d2b.png)

**Comment by barryvdh** [comment]
I'm having the same problem with 99% discount rates.

**Comment by HenKun** [comment]
@barryvdh Did you try the patch proposed by https://github.com/magento/magento2/issues/30853#issuecomment-764582933 ?

**Comment by HenKun** [comment]
The patch in https://github.com/magento/magento2/issues/30853#issuecomment-764582933 could not be applied, so I created it on my own:

```
diff --git a/Model/Order/Invoice/Total/Discount.php b/Model/Order/Invoice/Total/Discount.php
index ef7205b..0807c5d 100644
--- a/Model/Order/Invoice/Total/Discount.php
+++ b/Model/Order/Invoice/Total/Discount.php
@@ -71,12 +71,11 @@ class Discount extends AbstractTotal
         $invoice->setDiscountAmount(-$totalDiscountAmount);
         $invoice->setBaseDiscountAmount(-$baseTotalDiscountAmount);

-        $grandTotal = $invoice->getGrandTotal() - $totalDiscountAmount < 0.0001
-            ? 0 : $invoice->getGrandTotal() - $totalDiscountAmount;
-        $baseGrandTotal = $invoice->getBaseGrandTotal() - $baseTotalDiscountAmount < 0.0001
-            ? 0 : $invoice->getBaseGrandTotal() - $baseTotalDiscountAmount;
-        $invoice->setGrandTotal($grandTotal);
-        $invoice->setBaseGrandTotal($baseGrandTotal);
+        // Lines reverted from 2.3.5 as https://github.com/magento/magento2/commit/c710c9b38e0bcfbaedc6114bc5b4b9e62df95bbf
+        // bogs it up. Ref: https://github.com/magento/magento2/issues/30853
+
+        $invoice->setGrandTotal($invoice->getGrandTotal() - $totalDiscountAmount);
+        $invoice->setBaseGrandTotal($invoice->getBaseGrandTotal() - $baseTotalDiscountAmount);
         return $this;
     }

```

I tested it with discounts of 0%, 40%, 50%, 90% and 100%

All tests worked fine with my setting. (On M2.4.1)

However, I am not sure what the commit  https://github.com/magento/magento2/commit/c710c9b38e0bcfbaedc6114bc5b4b9e62df95bbf  `MC-35554: Orders with Zero Payment Information required are Closed after being invoiced` that introduced the bug wanted to achieve with this change, so it might have other side effects reverting this change with this patch.

**Comment by hostep** [comment]
This is probably fixed by: [MC-41438: 100% Discount cart rule results in Tax invoiced](https://github.com/magento/magento2/commit/533ec2eb46c91b79b7a99ba4a6a164e2f6e3d7f1), which I'm guessing is going to be included in Magento 2.4.3

Maybe somebody can confirm this?

**Comment by engcom-Alfa** [comment]
Hi @HenKun 

I have re tested it in the Magento ver. dev-2.4-develop, and it is worked as expected. Below are the screenshots of it.

![image](https://user-images.githubusercontent.com/51679138/119494266-bda26e80-bd7e-11eb-90ac-3a745ff18b1c.png)

![image](https://user-images.githubusercontent.com/51679138/119494302-c98e3080-bd7e-11eb-83ff-f5e5536dfa12.png)

![image](https://user-images.githubusercontent.com/51679138/119494343-d6128900-bd7e-11eb-97a2-8e6ccc55ae15.png)

![image](https://user-images.githubusercontent.com/51679138/119494391-e460a500-bd7e-11eb-962e-ab2365456b11.png)

![image](https://user-images.githubusercontent.com/51679138/119494449-f0e4fd80-bd7e-11eb-9fc7-f1af63b8b2b2.png)

So, kindly update to the latest version and give a try. If you still face the issue, you may please report a new bug on the same.

**Comment by vrielsa** [comment]
The fix on develop and the one mentioned here doesn't work if your discount does not apply to shipping costs.
![Screenshot 2021-07-08 at 13 42 06](https://user-images.githubusercontent.com/33426821/124917151-c5973480-dff3-11eb-8fd0-59d4c8400dca.png)

**Comment by valguss** [comment]
> This is probably fixed by: [MC-41438: 100% Discount cart rule results in Tax invoiced](https://github.com/magento/magento2/commit/533ec2eb46c91b79b7a99ba4a6a164e2f6e3d7f1), which I'm guessing is going to be included in Magento 2.4.3
> 
> Maybe somebody can confirm this?

Yeah, just had to apply the patch you mentioned and that has fixed the issue for me
